### PR TITLE
fix: audit fixes

### PR DESCRIPTION
**Pull request by oren-lava**
# Description

Closes: #XXXX

In this PR I fixed a few issues that came up during an audit of the consensus code.

1. Added a warning comment on proper use of `DetectionIndex()`

2. Added safety checks to avoid potential nil dereference in the node's code. The potential nil dereference issues were found by Uber's nilaway tool ([link](https://github.com/uber-go/nilaway)).
nilaway's original reports are attached. Note, I did not fix potential errors in test code.

3. Removed the pairing query cache.
---
[nilaway_reports.zip](https://github.com/user-attachments/files/16868993/nilaway_reports.zip)

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] read the [contribution guide](https://github.com/lavanet/lava/blob/main/CONTRIBUTING.md)
* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the `main` branch
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration tests
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: x/conflict/keeper/conflict.go [hunk 1]
```diff
@@ -20,6 +20,11 @@ func (k Keeper) ValidateFinalizationConflict(ctx sdk.Context, conflictData *type
 }
 
 func (k Keeper) ValidateResponseConflict(ctx sdk.Context, conflictData *types.ResponseConflict, clientAddr sdk.AccAddress) error {
+	// 0. validate conflictData is not nil
+	if conflictData.IsDataNil() {
```
Comment (Yaroms) [inline x/conflict/keeper/conflict.go:24]: can this be also in the validate basic?
also if we check it here we dont have to check it anymore, the rest are redundant

File: x/conflict/keeper/msg_server_detection.go [hunk 2]
```diff
@@ -14,7 +14,12 @@ import (
 	"golang.org/x/exp/slices"
 )
 
+// DetectionIndex creates an index for detection instances.
+// WARNING: the detection index should not be used for prefixed iteration.
 func DetectionIndex(creatorAddr string, conflict *types.ResponseConflict, epochStart uint64) string {
+	if conflict.IsDataNil() {
```
Comment (Yaroms) [inline x/conflict/keeper/msg_server_detection.go:20]: redundant

File: x/pairing/keeper/scores/geo_req.go [hunk 3]
```diff
@@ -20,13 +20,17 @@ const (
 	minGeoLatency        = 1
 )
 
-func (gr GeoReq) Init(policy planstypes.Policy) bool {
-	return true
+func (gr *GeoReq) Init(policy planstypes.Policy) bool {
+	return gr != nil
```
Comment (Yaroms) [inline x/pairing/keeper/scores/geo_req.go:24]: is this even possible?
can gr be nil?
its confusing for nothing

File: x/pairing/keeper/scores/geo_req.go [hunk 4]
```diff
@@ -20,13 +20,17 @@ const (
...
 }
 
 // Score calculates the geo score of a provider based on preset latency data
 // Note: each GeoReq must have exactly a single geolocation (bit)
-func (gr GeoReq) Score(score PairingScore) math.Uint {
+func (gr *GeoReq) Score(score PairingScore) math.Uint {
+	if gr == nil {
```
Comment (Yaroms) [inline x/pairing/keeper/scores/geo_req.go:30]: same?

File: x/pairing/keeper/scores/geo_req.go [hunk 5]
```diff
@@ -38,13 +42,16 @@ func (gr GeoReq) Score(score PairingScore) math.Uint {
 	return cost
 }
 
-func (gr GeoReq) GetName() string {
+func (gr *GeoReq) GetName() string {
```
Comment (Yaroms) [inline x/pairing/keeper/scores/geo_req.go:45]: why do we make it pointers?

File: x/projects/types/project.go [hunk 6]
```diff
@@ -82,6 +82,9 @@ func (project *Project) GetKey(key string) ProjectKey {
 }
 
 func (project *Project) AppendKey(key ProjectKey) bool {
+	if project == nil {
```
Comment (Yaroms) [inline x/projects/types/project.go:85]: can never be nil

File: x/conflict/keeper/msg_server_detection.go [hunk 7]
```diff
@@ -125,6 +130,11 @@ func (k msgServer) handleSameProviderFinalizationConflict(ctx sdk.Context, confl
 }
 
 func (k msgServer) handleResponseConflict(ctx sdk.Context, goCtx context.Context, conflict *types.ResponseConflict, clientAddr sdk.AccAddress) (eventData map[string]string, err error) {
+	if conflict.IsDataNil() {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/conflict/keeper/msg_server_detection.go]: done. see https://github.com/lavanet/lava/commit/bfaa18c0969b329cfb0d19481615c5f0cec66cfe

File: x/conflict/keeper/msg_server_detection.go (hunk 2 above)
Comment (oren-lava) [reply x/conflict/keeper/msg_server_detection.go]: done. see https://github.com/lavanet/lava/commit/bfaa18c0969b329cfb0d19481615c5f0cec66cfe

File: x/conflict/keeper/conflict.go (hunk 1 above)
Comment (oren-lava) [reply x/conflict/keeper/conflict.go:24]: done. see https://github.com/lavanet/lava/commit/bfaa18c0969b329cfb0d19481615c5f0cec66cfe

File: x/pairing/keeper/scores/geo_req.go (hunk 5 above)
Comment (oren-lava) [reply x/pairing/keeper/scores/geo_req.go]: reverted. see https://github.com/lavanet/lava/commit/b2c155b44d7902f6ba4371cdbcb58ad1d51caea6

**Review by Yaroms — DISMISSED** [review]

File: x/conflict/keeper/msg_server_detection.go [hunk 8]
```diff
@@ -14,6 +14,8 @@ import (
 	"golang.org/x/exp/slices"
 )
 
+// DetectionIndex creates an index for detection instances.
+// WARNING: the detection index should not be used for prefixed iteration.
```
Comment (omerlavanet) [inline x/conflict/keeper/msg_server_detection.go:18]: add why please

**Review by omerlavanet — DISMISSED** [review]

File: x/conflict/keeper/msg_server_detection.go (hunk 8 above)
Comment (oren-lava) [reply x/conflict/keeper/msg_server_detection.go]: done. see https://github.com/lavanet/lava/commit/9218823d827a66b0faf97dd88a5b3a32be1b11cf

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**